File: packages/client/src/namespaces/mcp.ts
//! `mcp.*` client namespace. Wave 2 owner: D (M1.11).
//!
//! TODO(D): type `health` against tethys-api; it is still `call<void>()`.

import type {
  Applied,
  AttachmentGrid,
  ImportCandidate,
  ImportScan,
  ProjectionPlan,
  RegistryEntry,
  RegistryEntryView,
  Scope,
  TargetId,
  VerifyStatus,
} from "@tethys/bindings";

import type { Call } from "../transport";

export function mcpNamespace(call: Call) {
  return {
    registryList: (workspaceId?: string) =>
      call<RegistryEntryView[]>("mcp_registry_list", { workspaceId }),
    registry_list: (workspaceId?: string) =>
      call<RegistryEntryView[]>("mcp_registry_list", { workspaceId }),
    registrySet: (
      name: string,
      entry: RegistryEntry,
      scope: Scope,
      workspaceId?: string,
    ) => call<void>("mcp_registry_set", { name, entry, scope, workspaceId }),
    registry_set: (
      name: string,
      entry: RegistryEntry,
      scope: Scope,
      workspaceId?: string,
    ) => call<void>("mcp_registry_set", { name, entry, scope, workspaceId }),
    registryDelete: (name: string, scope: Scope, workspaceId?: string) =>
      call<boolean>("mcp_registry_delete", { name, scope, workspaceId }),
    registry_delete: (name: string, scope: Scope, workspaceId?: string) =>
      call<boolean>("mcp_registry_delete", { name, scope, workspaceId }),
    effective: (providerId?: string, workspaceId?: string) =>
      call<RegistryEntryView[]>("mcp_effective", { providerId, workspaceId }),
    attachments: (workspaceId: string) =>
      call<AttachmentGrid>("mcp_attachments", { workspaceId }),
    projectionPlan: (workspaceId: string, target: TargetId, scope: Scope) =>
      call<ProjectionPlan>("mcp_projection_plan", {
        workspaceId,
        target,
        scope,
      }),
    projection_plan: (workspaceId: string, target: TargetId, scope: Scope) =>
      call<ProjectionPlan>("mcp_projection_plan", {
        workspaceId,
        target,
        scope,
      }),
    projectionApply: (
      workspaceId: string,
      target: TargetId,
      scope: Scope,
      plan: ProjectionPlan,
    ) =>
      call<Applied>("mcp_projection_apply", {
        workspaceId,
        target,
        scope,
        plan,
      }),
    projection_apply: (
      workspaceId: string,
      target: TargetId,
      scope: Scope,
      plan: ProjectionPlan,
    ) =>
      call<Applied>("mcp_projection_apply", {
        workspaceId,
        target,
        scope,
        plan,
      }),
    projectionRollback: (workspaceId: string, target: TargetId, scope: Scope) =>
      call<void>("mcp_projection_rollback", { workspaceId, target, scope }),
    projection_rollback: (
      workspaceId: string,
      target: TargetId,
      scope: Scope,
    ) => call<void>("mcp_projection_rollback", { workspaceId, target, scope }),
    projectionVerify: (workspaceId: string, target: TargetId, scope: Scope) =>
      call<VerifyStatus>("mcp_projection_verify", {
        workspaceId,
        target,
        scope,
      }),
    projection_verify: (workspaceId: string, target: TargetId, scope: Scope) =>
      call<VerifyStatus>("mcp_projection_verify", {
        workspaceId,
        target,
        scope,
      }),
    importScan: (workspaceId: string) =>
      call<ImportScan>("mcp_import_scan", { workspaceId }),
    import_scan: (workspaceId: string) =>
      call<ImportScan>("mcp_import_scan", { workspaceId }),
    importApply: (
      workspaceId: string,
      candidates: ImportCandidate[],
      scope: Scope,
    ) =>
      call<string[]>("mcp_import_apply", {
        workspaceId,
        candidates,
        scope,
      }),
    import_apply: (
      workspaceId: string,
      candidates: ImportCandidate[],
      scope: Scope,
    ) =>
      call<string[]>("mcp_import_apply", {
        workspaceId,
        candidates,
        scope,
      }),
    health: () => call<void>("mcp_health"),
  };
}

File: crates/tethys-api/src/mcp.rs
//! `mcp.*` namespace (`architecture.md` §12.1).

use tethys_schema::sync::{
    Applied, AttachmentGrid, ImportCandidate, ImportScan, ProjectionPlan, RegistryEntry,
    RegistryEntryView, Scope, TargetId, VerifyStatus, WorkspaceId,
};

use crate::ApiError;

/// Registry, projection, and import methods.
pub trait McpApi: Send + Sync {
    fn mcp_registry_list(
        &self,
        _workspace_id: Option<WorkspaceId>,
    ) -> impl std::future::Future<Output = Result<Vec<RegistryEntryView>, ApiError>> + Send {
        async { Err(ApiError::Unimplemented("mcp.registry_list")) }
    }

    fn mcp_registry_set(
        &self,
        _name: String,
        _entry: RegistryEntry,
        _scope: Scope,
        _workspace_id: Option<WorkspaceId>,
    ) -> impl std::future::Future<Output = Result<(), ApiError>> + Send {
        async { Err(ApiError::Unimplemented("mcp.registry_set")) }
    }

    fn mcp_registry_delete(
        &self,
        _name: String,
        _scope: Scope,
        _workspace_id: Option<WorkspaceId>,
    ) -> impl std::future::Future<Output = Result<bool, ApiError>> + Send {
        async { Err(ApiError::Unimplemented("mcp.registry_delete")) }
    }

    fn mcp_effective(
        &self,
        _provider_id: Option<String>,
        _workspace_id: Option<WorkspaceId>,
    ) -> impl std::future::Future<Output = Result<Vec<RegistryEntryView>, ApiError>> + Send {
        async { Err(ApiError::Unimplemented("mcp.effective")) }
    }

    fn mcp_attachments(
        &self,
        _workspace_id: WorkspaceId,
    ) -> impl std::future::Future<Output = Result<AttachmentGrid, ApiError>> + Send {
        async { Err(ApiError::Unimplemented("mcp.attachments")) }
    }

    fn mcp_projection_plan(
        &self,
        _workspace_id: WorkspaceId,
        _target: TargetId,
        _scope: Scope,
    ) -> impl std::future::Future<Output = Result<ProjectionPlan, ApiError>> + Send {
        async { Err(ApiError::Unimplemented("mcp.projection_plan")) }
    }

    fn mcp_projection_apply(
        &self,
        _workspace_id: WorkspaceId,
        _target: TargetId,
        _scope: Scope,
        _plan: ProjectionPlan,
    ) -> impl std::future::Future<Output = Result<Applied, ApiError>> + Send {
        async { Err(ApiError::Unimplemented("mcp.projection_apply")) }
    }

    fn mcp_projection_rollback(
        &self,
        _workspace_id: WorkspaceId,
        _target: TargetId,
        _scope: Scope,
    ) -> impl std::future::Future<Output = Result<(), ApiError>> + Send {
        async { Err(ApiError::Unimplemented("mcp.projection_rollback")) }
    }

    fn mcp_projection_verify(
        &self,
        _workspace_id: WorkspaceId,
        _target: TargetId,
        _scope: Scope,
    ) -> impl std::future::Future<Output = Result<VerifyStatus, ApiError>> + Send {
        async { Err(ApiError::Unimplemented("mcp.projection_verify")) }
    }

    fn mcp_import_scan(
        &self,
        _workspace_id: WorkspaceId,
    ) -> impl std::future::Future<Output = Result<ImportScan, ApiError>> + Send {
        async { Err(ApiError::Unimplemented("mcp.import_scan")) }
    }

    fn mcp_import_apply(
        &self,
        _workspace_id: WorkspaceId,
        _candidates: Vec<ImportCandidate>,
        _scope: Scope,
    ) -> impl std::future::Future<Output = Result<Vec<String>, ApiError>> + Send {
        async { Err(ApiError::Unimplemented("mcp.import_apply")) }
    }

    fn mcp_health(&self) -> impl std::future::Future<Output = Result<(), ApiError>> + Send {
        async { Err(ApiError::Unimplemented("mcp.health")) }
    }
}
